vm/src/stdlib/marshal.rs

```diff
@@ -237,158 +196,172 @@ mod decl {
+        }
+        Ok(vec)
     }
 
     #[pyfunction]
-    fn loads(pybuffer: PyBuffer, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
-        let buf = pybuffer.as_contiguous().ok_or_else(|| {
-            vm.new_buffer_error("Buffer provided to marshal.loads() is not contiguous".to_owned())
-        })?;
-        let (obj, _) = load_obj(&buf, vm)?;
-        Ok(obj)
+    fn loads(pybuffer: PyBuffer, vm: &VirtualMachine) -> MarshalReadResult<PyObjectRef> {
+        let buf = pybuffer
+            .as_contiguous()
+            .ok_or(MarshalReadError::BufferDiscontiguous)?;
+        load_obj(&mut ReadBuf(&buf), vm)
     }
 
-    fn load_obj<'b>(buf: &'b [u8], vm: &VirtualMachine) -> PyResult<(PyObjectRef, &'b [u8])> {
-        let (type_indicator, buf) = buf.split_first().ok_or_else(|| too_short_error(vm))?;
-        let typ = Type::try_from(*type_indicator)
-            .map_err(|_| vm.new_value_error("bad marshal data (unknown type code)".to_owned()))?;
-        let (obj, buf) = match typ {
-            Type::True => (true.to_pyobject(vm), buf),
-            Type::False => (false.to_pyobject(vm), buf),
-            Type::None => (vm.ctx.none(), buf),
-            Type::Ellipsis => (vm.ctx.ellipsis(), buf),
+    fn load_obj(buf: &mut ReadBuf<'_>, vm: &VirtualMachine) -> MarshalReadResult<PyObjectRef> {
+        let [type_indicator] = *buf.read_n_const::<1>()?;
+        let typ = Type::try_from(type_indicator).map_err(|_| MarshalReadError::UnknownType)?;
+        let obj = match typ {
+            Type::True => true.to_pyobject(vm),
+            Type::False => false.to_pyobject(vm),
+            Type::None => vm.ctx.none(),
+            Type::Ellipsis => vm.ctx.ellipsis(),
             Type::Int => {
-                if buf.len() < 4 {
-                    return Err(too_short_error(vm));
-                }
-                let (len_bytes, buf) = buf.split_at(4);
-                let len = i32::from_le_bytes(len_bytes.try_into().unwrap());
-                let (sign, len) = if len < 0 {
-                    (Sign::Minus, (-len) as usize)
-                } else {
-                    (Sign::Plus, len as usize)
-                };
-                if buf.len() < len {
-                    return Err(too_short_error(vm));
-                }
-                let (bytes, buf) = buf.split_at(len);
-                let int = BigInt::from_bytes_le(sign, bytes);
-                (int.to_pyobject(vm), buf)
-            }
-            Type::Float => {
-                if buf.len() < 8 {
-                    return Err(too_short_error(vm));
-                }
-                let (bytes, buf) = buf.split_at(8);
-                let number = f64::from_le_bytes(bytes.try_into().unwrap());
-                (vm.ctx.new_float(number).into(), buf)
+                let len = i32::from_le_bytes(*buf.read_n_const::<4>()?);
+                let sign = if len < 0 { Sign::Minus } else { Sign::Plus };
+                let len = len.unsigned_abs() as usize;
```

Comment:
Not a big deal but this is changed from single comparison to double comparison (symentically)

```suggestion
                let (sign, len) = if len < 0 {
                    (Sign::Minus, (-len) as usize)
                } else {
                    (Sign::Plus, len as usize)
                };
```

Reply:
I actually prefer what @coolreader18 wrote here because its intention is clearer. However, I do have one improvement for the first line:

```suggestion
                let sign = if len.is_negative() { Sign::Minus } else { Sign::Plus };
                let len = len.unsigned_abs() as usize;
```